### ISSUE TITLE
with gopls v0.6.4, eglot-code-action-organize-imports fails due to version skew when doing lots of stuff in before-save-hook

### ISSUE THREAD

**Issue by bcmills** [joaotavora/eglot#609]
### Summary

`eglot-code-action-organize-imports` sometimes fails if run immediately after an edit to the buffer.

### `*Messages*`

```
Loading /etc/google-emacs/site-start.d/00debian.el (source)...done
Loading /etc/google-emacs/site-start.d/50emacs-google-config.el (source)...done
For information about GNU Emacs and the GNU system, type C-h C-a.
[yas] Prepared just-in-time loading of snippets (but no snippets found).
Loading eldoc...done
Saving file /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go...
[eglot] Server reports (type=4): Loading packages...
[eglot] Server reports (type=3): Finished loading packages.
Error: (jsonrpc-error "Edits on ‘main.go’ require version 0, you have 1" (jsonrpc-error-code . 32603) (jsonrpc-error-message . "Edits on ‘main.go’ require version 0, you have 1"))
Wrote /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go
[eglot] Connected! Server `gopls' now managing `go-mode' buffers in project `eglot-issue-411'.
Package cl is deprecated
Quit
```

### `M-x eglot-events-buffer`

```
client-request (id:1) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :id 1 :method "initialize" :params
	  (:processId 2609622 :rootPath "/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :rootUri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :initializationOptions #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
																													       ())
		      :capabilities
		      (:workspace
		       (:applyEdit t :executeCommand
				   (:dynamicRegistration :json-false)
				   :workspaceEdit
				   (:documentChanges :json-false)
				   :didChangeWatchedFiles
				   (:dynamicRegistration t)
				   :symbol
				   (:dynamicRegistration :json-false)
				   :configuration t)
		       :textDocument
		       (:synchronization
			(:dynamicRegistration :json-false :willSave t :willSaveWaitUntil t :didSave t)
			:completion
			(:dynamicRegistration :json-false :completionItem
					      (:snippetSupport t)
					      :contextSupport t)
			:hover
			(:dynamicRegistration :json-false :contentFormat
					      ["markdown" "plaintext"])
			:signatureHelp
			(:dynamicRegistration :json-false :signatureInformation
					      (:parameterInformation
					       (:labelOffsetSupport t)))
			:references
			(:dynamicRegistration :json-false)
			:definition
			(:dynamicRegistration :json-false)
			:declaration
			(:dynamicRegistration :json-false)
			:implementation
			(:dynamicRegistration :json-false)
			:typeDefinition
			(:dynamicRegistration :json-false)
			:documentSymbol
			(:dynamicRegistration :json-false :hierarchicalDocumentSymbolSupport t :symbolKind
					      (:valueSet
					       [1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26]))
			:documentHighlight
			(:dynamicRegistration :json-false)
			:codeAction
			(:dynamicRegistration :json-false :codeActionLiteralSupport
					      (:codeActionKind
					       (:valueSet
						["quickfix" "refactor" "refactor.extract" "refactor.inline" "refactor.rewrite" "source" "source.organizeImports"]))
					      :isPreferredSupport t)
			:formatting
			(:dynamicRegistration :json-false)
			:rangeFormatting
			(:dynamicRegistration :json-false)
			:rename
			(:dynamicRegistration :json-false)
			:publishDiagnostics
			(:relatedInformation :json-false))
		       :experimental #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
						   ()))))

server-reply (id:1) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :result
	  (:capabilities
	   (:textDocumentSync
	    (:openClose t :change 2 :save nil)
	    :completionProvider
	    (:triggerCharacters
	     ["."])
	    :hoverProvider t :signatureHelpProvider
	    (:triggerCharacters
	     ["(" ","])
	    :definitionProvider t :typeDefinitionProvider t :implementationProvider t :referencesProvider t :documentHighlightProvider t :documentSymbolProvider t :codeActionProvider
	    (:codeActionKinds
	     ["quickfix" "refactor.extract" "refactor.rewrite" "source.fixAll" "source.organizeImports"])
	    :codeLensProvider nil :documentLinkProvider nil :workspaceSymbolProvider t :documentFormattingProvider t :documentOnTypeFormattingProvider
	    (:firstTriggerCharacter "")
	    :renameProvider t :foldingRangeProvider t :executeCommandProvider
	    (:commands
	     ["gopls.generate" "gopls.fill_struct" "gopls.regenerate_cgo" "gopls.test" "gopls.tidy" "gopls.update_go_sum" "gopls.undeclared_name" "gopls.go_get_package" "gopls.add_dependency" "gopls.upgrade_dependency" "gopls.remove_dependency" "gopls.vendor" "gopls.extract_variable" "gopls.extract_function" "gopls.gc_details" "gopls.generate_gopls_mod"])
	    :callHierarchyProvider t :workspace
	    (:workspaceFolders
	     (:supported t :changeNotifications "workspace/didChangeWorkspaceFolders")))
	   :serverInfo
	   (:name "gopls" :version "{\"path\":\"golang.org/x/tools/gopls\",\"version\":\"v0.6.4\",\"sum\":\"h1:PO8rURGmi+QEKZ3pk+ZXJh51EFTVTSXqxzoRzrGSTJM=\",\"deps\":[{\"path\":\"github.com/BurntSushi/toml\",\"version\":\"v0.3.1\",\"sum\":\"h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=\"},{\"path\":\"github.com/google/go-cmp\",\"version\":\"v0.5.4\",\"sum\":\"h1:L8R9j+yAqZuZjsqh/z+F1NCffTKKLShY6zXTItVIZ8M=\"},{\"path\":\"github.com/sergi/go-diff\",\"version\":\"v1.1.0\",\"sum\":\"h1:we8PVUC3FE2uYfodKH/nBHMSetSfHDR6scGdBi+erh0=\"},{\"path\":\"golang.org/x/mod\",\"version\":\"v0.4.0\",\"sum\":\"h1:8pl+sMODzuvGJkmj2W4kZihvVb5mKm8pB/X44PIQHv8=\"},{\"path\":\"golang.org/x/sync\",\"version\":\"v0.0.0-20201020160332-67f06af15bc9\",\"sum\":\"h1:SQFwaSi55rU7vdNs9Yr0Z324VNlrF+0wMqRXT4St8ck=\"},{\"path\":\"golang.org/x/sys\",\"version\":\"v0.0.0-20210119212857-b64e53b001e4\",\"sum\":\"h1:myAQVi0cGEoqQVR5POX+8RR2mrocKqNN1hmeMqhX27k=\"},{\"path\":\"golang.org/x/tools\",\"version\":\"v0.1.1-0.20210119222907-0a1a9685734a\",\"sum\":\"h1:ttr/sghFptib3jlHwNS65KwGoqjr03TFzso3kpHyia8=\"},{\"path\":\"golang.org/x/xerrors\",\"version\":\"v0.0.0-20200804184101-5ec99f83aff1\",\"sum\":\"h1:go1bK/D/BFZV2I8cIQd1NKEZ+0owSTG1fDTci4IqFcE=\"},{\"path\":\"honnef.co/go/tools\",\"version\":\"v0.0.1-2020.1.6\",\"sum\":\"h1:W18jzjh8mfPez+AwGLxmOImucz/IFjpNlrKVnaj2YVc=\"},{\"path\":\"mvdan.cc/gofumpt\",\"version\":\"v0.1.0\",\"sum\":\"h1:hsVv+Y9UsZ/mFZTxJZuHVI6shSQCtzZ11h1JEFPAZLw=\"},{\"path\":\"mvdan.cc/xurls/v2\",\"version\":\"v2.2.0\",\"sum\":\"h1:NSZPykBXJFCetGZykLAxaL6SIpvbVy/UFEniIfHAa8A=\"}]}"))
	  :id 1)

client-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "initialized" :params #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
							    ()))

client-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "textDocument/didOpen" :params
	  (:textDocument
	   (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 0 :languageId "go" :text "  package main\n\nfunc main() {\n	  type Version = module.Version\n}\n")))

client-request (id:2) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :id 2 :method "textDocument/codeAction" :params
	  (:textDocument
	   (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
	   :range
	   (:start
	    (:line 0 :character 1)
	    :end
	    (:line 0 :character 1))
	   :context
	   (:diagnostics
	    []
	    :only
	    ["source.organizeImports"])))

server-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "window/showMessage" :params
	  (:type 4 :message "Loading packages..."))

server-request (id:1) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "workspace/configuration" :params
	  (:items
	   [(:scopeUri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :section "gopls")])
	  :id 1)

client-reply (id:1) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :id 1 :result
	  [nil])

server-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
	  (:type 3 :message "2021/01/29 12:35:04 go env for /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/\n(root /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411)\n(go version go version devel +1de5e22b63 Thu Jan 28 10:07:48 2021 -0500 linux/amd64)\n(valid build configuration = true)\n(build flags: [])\nGOPATH=/usr/local/google/home/bcmills\nGOINSECURE=\nGOROOT=/usr/local/google/home/bcmills/go\nGOCACHE=/usr/local/google/home/bcmills/.cache/go-build\nGOMOD=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod\nGONOPROXY=\nGOPRIVATE=\nGOPROXY=https://proxy.golang.org,direct\nGO111MODULE=\nGOFLAGS=-benchtime=1x\nGOMODCACHE=/usr/local/google/home/bcmills/pkg/mod\nGONOSUMDB=\nGOSUMDB=sum.golang.org\n\n"))

server-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
	  (:type 3 :message "2021/01/29 12:35:04 go/packages.Load\n	snapshot=0\n	directory=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411\n	query=[builtin example.com/...]\n	packages=2\n"))

server-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "window/showMessage" :params
	  (:type 3 :message "Finished loading packages."))

server-request (id:2) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "client/registerCapability" :params
	  (:registrations
	   [(:id "workspace/didChangeWatchedFiles-0" :method "workspace/didChangeWatchedFiles" :registerOptions
		 (:watchers
		  [(:globPattern "**/*.{go,mod,sum}" :kind 7)
		   (:globPattern "/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/**/*.{go,mod,sum}" :kind 7)]))])
	  :id 2)

client-reply (id:2) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :id 2 :result nil)

server-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
	  (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod" :diagnostics
		[(:range
		  (:start
		   (:line 4 :character 0)
		   :end
		   (:line 4 :character 31))
		  :severity 2 :source "go mod tidy" :message "golang.org/x/mod is not used in this module")]))

internal Fri Jan 29 12:35:04 2021:
(:message "Diagnostics received for unvisited (file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod)")

server-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
	  (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :diagnostics
		[(:range
		  (:start
		   (:line 3 :character 18)
		   :end
		   (:line 3 :character 24))
		  :severity 1 :source "compiler" :message "undeclared name: module")]))

server-reply (id:2) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :result
	  [(:title "Organize Imports" :kind "source.organizeImports" :disabled
		   (:reason "")
		   :edit
		   (:documentChanges
		    [(:textDocument
		      (:version 0 :uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
		      :edits
		      [(:range
			(:start
			 (:line 0 :character 0)
			 :end
			 (:line 0 :character 2))
			:newText "")
		       (:range
			(:start
			 (:line 1 :character 0)
			 :end
			 (:line 1 :character 0))
			:newText "\nimport \"golang.org/x/mod/module\"\n")])]))]
	  :id 2)

client-request (id:3) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :id 3 :method "textDocument/documentSymbol" :params
	  (:textDocument
	   (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")))

server-reply (id:3) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :result
	  [(:name "main" :detail "()" :kind 12 :range
		  (:start
		   (:line 2 :character 0)
		   :end
		   (:line 4 :character 1))
		  :selectionRange
		  (:start
		   (:line 2 :character 5)
		   :end
		   (:line 2 :character 9)))]
	  :id 3)

client-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "textDocument/didChange" :params
	  (:textDocument
	   (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 1)
	   :contentChanges
	   [(:range
	     (:start
	      (:line 0 :character 0)
	      :end
	      (:line 0 :character 0))
	     :rangeLength 0 :text " ")]))

client-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "textDocument/didSave" :params
	  (:text "   package main\n\nfunc main() {\n	  type Version = module.Version\n}\n" :textDocument
		 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")))

client-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
	  (:settings nil))

server-request (id:3) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "workspace/configuration" :params
	  (:items
	   [(:section "gopls")])
	  :id 3)

client-reply (id:3) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :id 3 :result
	  [nil])

server-request (id:4) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "workspace/configuration" :params
	  (:items
	   [(:scopeUri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :section "gopls")])
	  :id 4)

client-reply (id:4) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :id 4 :result
	  [nil])

```

<details>
<summary><code>gopls.log</code></summary>

```
[Trace - 12:35:04.322 PM] Sending request 'initialize - (1)'.
Params: {"processId":2609622,"rootPath":"/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/","rootUri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/","initializationOptions":{},"capabilities":{"workspace":{"applyEdit":true,"executeCommand":{"dynamicRegistration":false},"workspaceEdit":{"documentChanges":false},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":false},"configuration":true},"textDocument":{"synchronization":{"dynamicRegistration":false,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":false,"completionItem":{"snippetSupport":true},"contextSupport":true},"hover":{"dynamicRegistration":false,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":false,"signatureInformation":{"parameterInformation":{"labelOffsetSupport":true}}},"references":{"dynamicRegistration":false},"definition":{"dynamicRegistration":false},"declaration":{"dynamicRegistration":false},"implementation":{"dynamicRegistration":false},"typeDefinition":{"dynamicRegistration":false},"documentSymbol":{"dynamicRegistration":false,"hierarchicalDocumentSymbolSupport":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"documentHighlight":{"dynamicRegistration":false},"codeAction":{"dynamicRegistration":false,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}},"isPreferredSupport":true},"formatting":{"dynamicRegistration":false},"rangeFormatting":{"dynamicRegistration":false},"rename":{"dynamicRegistration":false},"publishDiagnostics":{"relatedInformation":false}},"experimental":{}}}


[Trace - 12:35:04.324 PM] Received response 'initialize - (1)' in 2ms.
Result: {"capabilities":{"textDocumentSync":{"openClose":true,"change":2,"save":{}},"completionProvider":{"triggerCharacters":["."]},"hoverProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",","]},"definitionProvider":true,"typeDefinitionProvider":true,"implementationProvider":true,"referencesProvider":true,"documentHighlightProvider":true,"documentSymbolProvider":true,"codeActionProvider":{"codeActionKinds":["quickfix","refactor.extract","refactor.rewrite","source.fixAll","source.organizeImports"]},"codeLensProvider":{},"documentLinkProvider":{},"workspaceSymbolProvider":true,"documentFormattingProvider":true,"documentOnTypeFormattingProvider":{"firstTriggerCharacter":""},"renameProvider":true,"foldingRangeProvider":true,"executeCommandProvider":{"commands":["gopls.generate","gopls.fill_struct","gopls.regenerate_cgo","gopls.test","gopls.tidy","gopls.update_go_sum","gopls.undeclared_name","gopls.go_get_package","gopls.add_dependency","gopls.upgrade_dependency","gopls.remove_dependency","gopls.vendor","gopls.extract_variable","gopls.extract_function","gopls.gc_details","gopls.generate_gopls_mod"]},"callHierarchyProvider":true,"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":"workspace/didChangeWorkspaceFolders"}}},"serverInfo":{"name":"gopls","version":"{\"path\":\"golang.org/x/tools/gopls\",\"version\":\"v0.6.4\",\"sum\":\"h1:PO8rURGmi+QEKZ3pk+ZXJh51EFTVTSXqxzoRzrGSTJM=\",\"deps\":[{\"path\":\"github.com/BurntSushi/toml\",\"version\":\"v0.3.1\",\"sum\":\"h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=\"},{\"path\":\"github.com/google/go-cmp\",\"version\":\"v0.5.4\",\"sum\":\"h1:L8R9j+yAqZuZjsqh/z+F1NCffTKKLShY6zXTItVIZ8M=\"},{\"path\":\"github.com/sergi/go-diff\",\"version\":\"v1.1.0\",\"sum\":\"h1:we8PVUC3FE2uYfodKH/nBHMSetSfHDR6scGdBi+erh0=\"},{\"path\":\"golang.org/x/mod\",\"version\":\"v0.4.0\",\"sum\":\"h1:8pl+sMODzuvGJkmj2W4kZihvVb5mKm8pB/X44PIQHv8=\"},{\"path\":\"golang.org/x/sync\",\"version\":\"v0.0.0-20201020160332-67f06af15bc9\",\"sum\":\"h1:SQFwaSi55rU7vdNs9Yr0Z324VNlrF+0wMqRXT4St8ck=\"},{\"path\":\"golang.org/x/sys\",\"version\":\"v0.0.0-20210119212857-b64e53b001e4\",\"sum\":\"h1:myAQVi0cGEoqQVR5POX+8RR2mrocKqNN1hmeMqhX27k=\"},{\"path\":\"golang.org/x/tools\",\"version\":\"v0.1.1-0.20210119222907-0a1a9685734a\",\"sum\":\"h1:ttr/sghFptib3jlHwNS65KwGoqjr03TFzso3kpHyia8=\"},{\"path\":\"golang.org/x/xerrors\",\"version\":\"v0.0.0-20200804184101-5ec99f83aff1\",\"sum\":\"h1:go1bK/D/BFZV2I8cIQd1NKEZ+0owSTG1fDTci4IqFcE=\"},{\"path\":\"honnef.co/go/tools\",\"version\":\"v0.0.1-2020.1.6\",\"sum\":\"h1:W18jzjh8mfPez+AwGLxmOImucz/IFjpNlrKVnaj2YVc=\"},{\"path\":\"mvdan.cc/gofumpt\",\"version\":\"v0.1.0\",\"sum\":\"h1:hsVv+Y9UsZ/mFZTxJZuHVI6shSQCtzZ11h1JEFPAZLw=\"},{\"path\":\"mvdan.cc/xurls/v2\",\"version\":\"v2.2.0\",\"sum\":\"h1:NSZPykBXJFCetGZykLAxaL6SIpvbVy/UFEniIfHAa8A=\"}]}"}}


[Trace - 12:35:04.325 PM] Sending notification 'initialized'.
Params: {}


[Trace - 12:35:04.325 PM] Received notification 'window/showMessage'.
Params: {"type":4,"message":"Loading packages..."}


[Trace - 12:35:04.326 PM] Received request 'workspace/configuration - (1)'.
Params: {"items":[{"scopeUri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/","section":"gopls"}]}


[Trace - 12:35:04.329 PM] Sending notification 'textDocument/didOpen'.
Params: {"textDocument":{"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go","version":0,"languageId":"go","text":"  package main\n\nfunc main() {\n\t  type Version = module.Version\n}\n"}}


[Trace - 12:35:04.331 PM] Sending request 'textDocument/codeAction - (2)'.
Params: {"textDocument":{"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go"},"range":{"start":{"line":0,"character":1},"end":{"line":0,"character":1}},"context":{"diagnostics":[],"only":["source.organizeImports"]}}


[Trace - 12:35:04.332 PM] Sending response 'workspace/configuration - (1)' in 6ms.
Result: [null]


[Trace - 12:35:04.392 PM] Received notification 'window/logMessage'.
Params: {"type":3,"message":"2021/01/29 12:35:04 go env for /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/\n(root /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411)\n(go version go version devel +1de5e22b63 Thu Jan 28 10:07:48 2021 -0500 linux/amd64)\n(valid build configuration = true)\n(build flags: [])\nGOPATH=/usr/local/google/home/bcmills\nGOINSECURE=\nGOROOT=/usr/local/google/home/bcmills/go\nGOCACHE=/usr/local/google/home/bcmills/.cache/go-build\nGOMOD=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod\nGONOPROXY=\nGOPRIVATE=\nGOPROXY=https://proxy.golang.org,direct\nGO111MODULE=\nGOFLAGS=-benchtime=1x\nGOMODCACHE=/usr/local/google/home/bcmills/pkg/mod\nGONOSUMDB=\nGOSUMDB=sum.golang.org\n\n"}


[Trace - 12:35:04.477 PM] Received notification 'window/logMessage'.
Params: {"type":3,"message":"2021/01/29 12:35:04 go/packages.Load\n\tsnapshot=0\n\tdirectory=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411\n\tquery=[builtin example.com/...]\n\tpackages=2\n"}


[Trace - 12:35:04.477 PM] Received notification 'window/showMessage'.
Params: {"type":3,"message":"Finished loading packages."}


[Trace - 12:35:04.478 PM] Received request 'client/registerCapability - (2)'.
Params: {"registrations":[{"id":"workspace/didChangeWatchedFiles-0","method":"workspace/didChangeWatchedFiles","registerOptions":{"watchers":[{"globPattern":"**/*.{go,mod,sum}","kind":7},{"globPattern":"/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/**/*.{go,mod,sum}","kind":7}]}}]}


[Trace - 12:35:04.486 PM] Received notification 'textDocument/publishDiagnostics'.
Params: {"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod","diagnostics":[{"range":{"start":{"line":4,"character":0},"end":{"line":4,"character":31}},"severity":2,"source":"go mod tidy","message":"golang.org/x/mod is not used in this module"}]}


[Trace - 12:35:04.486 PM] Received notification 'textDocument/publishDiagnostics'.
Params: {"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go","diagnostics":[{"range":{"start":{"line":3,"character":18},"end":{"line":3,"character":24}},"severity":1,"source":"compiler","message":"undeclared name: module"}]}


[Trace - 12:35:04.500 PM] Sending response 'client/registerCapability - (2)' in 22ms.
Result: 


[Trace - 12:35:04.907 PM] Received response 'textDocument/codeAction - (2)' in 575ms.
Result: [{"title":"Organize Imports","kind":"source.organizeImports","disabled":{"reason":""},"edit":{"documentChanges":[{"textDocument":{"version":0,"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go"},"edits":[{"range":{"start":{"line":0,"character":0},"end":{"line":0,"character":2}},"newText":""},{"range":{"start":{"line":1,"character":0},"end":{"line":1,"character":0}},"newText":"\nimport \"golang.org/x/mod/module\"\n"}]}]}}]


[Trace - 12:35:04.910 PM] Sending request 'textDocument/documentSymbol - (3)'.
Params: {"textDocument":{"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go"}}


[Trace - 12:35:04.911 PM] Received response 'textDocument/documentSymbol - (3)' in 0ms.
Result: [{"name":"main","detail":"()","kind":12,"range":{"start":{"line":2,"character":0},"end":{"line":4,"character":1}},"selectionRange":{"start":{"line":2,"character":5},"end":{"line":2,"character":9}}}]


[Trace - 12:35:04.968 PM] Sending notification 'textDocument/didChange'.
Params: {"textDocument":{"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go","version":1},"contentChanges":[{"range":{"start":{"line":0,"character":0},"end":{"line":0,"character":0}},"rangeLength":0,"text":" "}]}


[Trace - 12:35:04.969 PM] Sending notification 'textDocument/didSave'.
Params: {"text":"   package main\n\nfunc main() {\n\t  type Version = module.Version\n}\n","textDocument":{"uri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go"}}


[Trace - 12:35:04.969 PM] Sending notification 'workspace/didChangeConfiguration'.
Params: {"settings":null}


[Trace - 12:35:04.970 PM] Received request 'workspace/configuration - (3)'.
Params: {"items":[{"section":"gopls"}]}


[Trace - 12:35:04.971 PM] Sending response 'workspace/configuration - (3)' in 1ms.
Result: [null]


[Trace - 12:35:04.971 PM] Received request 'workspace/configuration - (4)'.
Params: {"items":[{"scopeUri":"file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/","section":"gopls"}]}


[Trace - 12:35:04.973 PM] Sending response 'workspace/configuration - (4)' in 1ms.
Result: [null]


[Trace - 12:35:34.908 PM] Received notification 'window/logMessage'.
Params: {"type":3,"message":"2021/01/29 12:35:34 background imports cache refresh starting\n"}


[Trace - 12:35:35.006 PM] Received notification 'window/logMessage'.
Params: {"type":3,"message":"2021/01/29 12:35:35 background refresh finished after 98.222662ms\n"}


```

</details>

### Steps to reproduce

0. Install and configure `gopls` [v0.6.4](https://github.com/golang/tools/tree/gopls/v0.6.4/gopls) and `eglot` at commit 2879b83.

1. Clone and check out https://github.com/bcmills/eglot-issue-411/commit/32e6e159f0da60432f79278ea37162fee4683cdb.

2. Either:
   a. In the repo root, run:
       ```
       emacs -q --load script.el
       ```
       or
   b. Bind a key to run `eglot-code-action-organize-imports` as described in https://github.com/joaotavora/eglot/issues/574#issuecomment-769645691. Then
       ```
       emacs main.go
       ```
       then add a leading space to the file and immediately press the bound key.

**Comment by joaotavora** [comment]
Let me just ask something. This is only when you tinker with `before-save-hook` right?  It's not when you call this directly, presumably...  If so, please amend the title to state that so it's clear.

And Eglot also uses `before-save-hook` so it might be a question of placing your hook before Eglot's . Or after them.   Else I will look at this later.

**Comment by bcmills** [comment]
> This is only when you tinker with `before-save-hook` right? It's not when you call this directly, presumably...

I'm pretty sure I've also seen it occur when running (a precursor to) `eglot-code-action-organize-imports` directly after a manual edit. It has to be pretty much immediate, though — and saving is about the only time I take an action like this immediately after an edit.

**Comment by joaotavora** [comment]
> I'm pretty sure I've also seen it occur when running (a precursor to) eglot-code-action-organize-imports directly after a manual edit. It has to be pretty much immediate, though 

Then it makes sense, because Eglot hasn't told the server about the changes yet.  It's not a use case I would worry much about (though it could be fixed).

What I don't understand is what kind of changes are happening when you save.  Saving doesn't change the buffer.

**Comment by joaotavora** [comment]
Oh, you're asking for the buffer to be formatted! And then immediately asking for code actions.

**Comment by bcmills** [comment]
Here is an event log from just adding some text, followed by triggering `eglot-code-action-organize-imports` from a keybinding.

```
[client-request] (id:1) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :id 1 :method "initialize" :params
					(:processId 2614217 :rootPath "/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :rootUri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :initializationOptions #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
																																																																																																																							 ())
											:capabilities
											(:workspace
											 (:applyEdit t :executeCommand
																	 (:dynamicRegistration :json-false)
																	 :workspaceEdit
																	 (:documentChanges :json-false)
																	 :didChangeWatchedFiles
																	 (:dynamicRegistration t)
																	 :symbol
																	 (:dynamicRegistration :json-false)
																	 :configuration t)
											 :textDocument
											 (:synchronization
												(:dynamicRegistration :json-false :willSave t :willSaveWaitUntil t :didSave t)
												:completion
												(:dynamicRegistration :json-false :completionItem
																							(:snippetSupport t)
																							:contextSupport t)
												:hover
												(:dynamicRegistration :json-false :contentFormat
																							["markdown" "plaintext"])
												:signatureHelp
												(:dynamicRegistration :json-false :signatureInformation
																							(:parameterInformation
																							 (:labelOffsetSupport t)))
												:references
												(:dynamicRegistration :json-false)
												:definition
												(:dynamicRegistration :json-false)
												:declaration
												(:dynamicRegistration :json-false)
												:implementation
												(:dynamicRegistration :json-false)
												:typeDefinition
												(:dynamicRegistration :json-false)
												:documentSymbol
												(:dynamicRegistration :json-false :hierarchicalDocumentSymbolSupport t :symbolKind
																							(:valueSet
																							 [1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26]))
												:documentHighlight
												(:dynamicRegistration :json-false)
												:codeAction
												(:dynamicRegistration :json-false :codeActionLiteralSupport
																							(:codeActionKind
																							 (:valueSet
																								["quickfix" "refactor" "refactor.extract" "refactor.inline" "refactor.rewrite" "source" "source.organizeImports"]))
																							:isPreferredSupport t)
												:formatting
												(:dynamicRegistration :json-false)
												:rangeFormatting
												(:dynamicRegistration :json-false)
												:rename
												(:dynamicRegistration :json-false)
												:publishDiagnostics
												(:relatedInformation :json-false))
											 :experimental #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
																									 ()))))
[server-reply] (id:1) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :result
					(:capabilities
					 (:textDocumentSync
						(:openClose t :change 2 :save nil)
						:completionProvider
						(:triggerCharacters
						 ["."])
						:hoverProvider t :signatureHelpProvider
						(:triggerCharacters
						 ["(" ","])
						:definitionProvider t :typeDefinitionProvider t :implementationProvider t :referencesProvider t :documentHighlightProvider t :documentSymbolProvider t :codeActionProvider
						(:codeActionKinds
						 ["quickfix" "refactor.extract" "refactor.rewrite" "source.fixAll" "source.organizeImports"])
						:codeLensProvider nil :documentLinkProvider nil :workspaceSymbolProvider t :documentFormattingProvider t :documentOnTypeFormattingProvider
						(:firstTriggerCharacter "")
						:renameProvider t :foldingRangeProvider t :executeCommandProvider
						(:commands
						 ["gopls.generate" "gopls.fill_struct" "gopls.regenerate_cgo" "gopls.test" "gopls.tidy" "gopls.update_go_sum" "gopls.undeclared_name" "gopls.go_get_package" "gopls.add_dependency" "gopls.upgrade_dependency" "gopls.remove_dependency" "gopls.vendor" "gopls.extract_variable" "gopls.extract_function" "gopls.gc_details" "gopls.generate_gopls_mod"])
						:callHierarchyProvider t :workspace
						(:workspaceFolders
						 (:supported t :changeNotifications "workspace/didChangeWorkspaceFolders")))
					 :serverInfo
					 (:name "gopls" :version "{\"path\":\"golang.org/x/tools/gopls\",\"version\":\"v0.6.4\",\"sum\":\"h1:PO8rURGmi+QEKZ3pk+ZXJh51EFTVTSXqxzoRzrGSTJM=\",\"deps\":[{\"path\":\"github.com/BurntSushi/toml\",\"version\":\"v0.3.1\",\"sum\":\"h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=\"},{\"path\":\"github.com/google/go-cmp\",\"version\":\"v0.5.4\",\"sum\":\"h1:L8R9j+yAqZuZjsqh/z+F1NCffTKKLShY6zXTItVIZ8M=\"},{\"path\":\"github.com/sergi/go-diff\",\"version\":\"v1.1.0\",\"sum\":\"h1:we8PVUC3FE2uYfodKH/nBHMSetSfHDR6scGdBi+erh0=\"},{\"path\":\"golang.org/x/mod\",\"version\":\"v0.4.0\",\"sum\":\"h1:8pl+sMODzuvGJkmj2W4kZihvVb5mKm8pB/X44PIQHv8=\"},{\"path\":\"golang.org/x/sync\",\"version\":\"v0.0.0-20201020160332-67f06af15bc9\",\"sum\":\"h1:SQFwaSi55rU7vdNs9Yr0Z324VNlrF+0wMqRXT4St8ck=\"},{\"path\":\"golang.org/x/sys\",\"version\":\"v0.0.0-20210119212857-b64e53b001e4\",\"sum\":\"h1:myAQVi0cGEoqQVR5POX+8RR2mrocKqNN1hmeMqhX27k=\"},{\"path\":\"golang.org/x/tools\",\"version\":\"v0.1.1-0.20210119222907-0a1a9685734a\",\"sum\":\"h1:ttr/sghFptib3jlHwNS65KwGoqjr03TFzso3kpHyia8=\"},{\"path\":\"golang.org/x/xerrors\",\"version\":\"v0.0.0-20200804184101-5ec99f83aff1\",\"sum\":\"h1:go1bK/D/BFZV2I8cIQd1NKEZ+0owSTG1fDTci4IqFcE=\"},{\"path\":\"honnef.co/go/tools\",\"version\":\"v0.0.1-2020.1.6\",\"sum\":\"h1:W18jzjh8mfPez+AwGLxmOImucz/IFjpNlrKVnaj2YVc=\"},{\"path\":\"mvdan.cc/gofumpt\",\"version\":\"v0.1.0\",\"sum\":\"h1:hsVv+Y9UsZ/mFZTxJZuHVI6shSQCtzZ11h1JEFPAZLw=\"},{\"path\":\"mvdan.cc/xurls/v2\",\"version\":\"v2.2.0\",\"sum\":\"h1:NSZPykBXJFCetGZykLAxaL6SIpvbVy/UFEniIfHAa8A=\"}]}"))
					:id 1)
[client-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "initialized" :params #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
																														()))
[client-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "textDocument/didOpen" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 0 :languageId "go" :text "package main\n\nfunc main() {\n	type Version = module.Version\n}\n")))
[client-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
					(:settings
					 (:gopls
						((staticcheck . t)
						 (matcher . "Fuzzy")))))
[client-request] (id:2) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :id 2 :method "textDocument/documentSymbol" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")))
[server-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "window/showMessage" :params
					(:type 4 :message "Loading packages..."))
[server-request] (id:1) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "workspace/configuration" :params
					(:items
					 [(:scopeUri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :section "gopls")])
					:id 1)
[client-reply] (id:1) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :id 1 :result
					[((staticcheck . t)
						(matcher . "Fuzzy"))])
[server-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
					(:type 3 :message "2021/01/29 13:06:00 go env for /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/\n(root /usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411)\n(go version go version devel +1de5e22b63 Thu Jan 28 10:07:48 2021 -0500 linux/amd64)\n(valid build configuration = true)\n(build flags: [])\nGOPROXY=https://proxy.golang.org,direct\nGOFLAGS=-benchtime=1x\nGOPRIVATE=\nGOROOT=/usr/local/google/home/bcmills/go\nGO111MODULE=\nGOMODCACHE=/usr/local/google/home/bcmills/pkg/mod\nGONOPROXY=\nGONOSUMDB=\nGOSUMDB=sum.golang.org\nGOCACHE=/usr/local/google/home/bcmills/.cache/go-build\nGOINSECURE=\nGOPATH=/usr/local/google/home/bcmills\nGOMOD=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod\n\n"))
[server-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
					(:type 3 :message "2021/01/29 13:06:00 go/packages.Load\n	snapshot=0\n	directory=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411\n	query=[builtin example.com/...]\n	packages=2\n"))
[server-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "window/showMessage" :params
					(:type 3 :message "Finished loading packages."))
[server-request] (id:2) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "client/registerCapability" :params
					(:registrations
					 [(:id "workspace/didChangeWatchedFiles-0" :method "workspace/didChangeWatchedFiles" :registerOptions
								 (:watchers
									[(:globPattern "**/*.{go,mod,sum}" :kind 7)
									 (:globPattern "/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/**/*.{go,mod,sum}" :kind 7)]))])
					:id 2)
[client-reply] (id:2) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :id 2 :result nil)
[server-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
					(:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :diagnostics
								[(:range
									(:start
									 (:line 3 :character 16)
									 :end
									 (:line 3 :character 22))
									:severity 1 :source "compiler" :message "undeclared name: module")]))
[server-notification] Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
					(:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod" :diagnostics
								[(:range
									(:start
									 (:line 4 :character 0)
									 :end
									 (:line 4 :character 31))
									:severity 2 :source "go mod tidy" :message "golang.org/x/mod is not used in this module")]))
[internal] Fri Jan 29 13:06:00 2021:
(:message "Diagnostics received for unvisited (file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/go.mod)")
[server-request] (id:3) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "workspace/configuration" :params
					(:items
					 [(:section "gopls")])
					:id 3)
[client-reply] (id:3) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :id 3 :result
					[((staticcheck . t)
						(matcher . "Fuzzy"))])
[server-request] (id:4) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :method "workspace/configuration" :params
					(:items
					 [(:scopeUri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/" :section "gopls")])
					:id 4)
[client-reply] (id:4) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :id 4 :result
					[((staticcheck . t)
						(matcher . "Fuzzy"))])
[server-reply] (id:2) Fri Jan 29 13:06:00 2021:
(:jsonrpc "2.0" :result
					[(:name "main" :detail "()" :kind 12 :range
									(:start
									 (:line 2 :character 0)
									 :end
									 (:line 4 :character 1))
									:selectionRange
									(:start
									 (:line 2 :character 5)
									 :end
									 (:line 2 :character 9)))]
					:id 2)
[client-request] (id:3) Fri Jan 29 13:06:02 2021:
(:jsonrpc "2.0" :id 3 :method "textDocument/codeAction" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
					 :range
					 (:start
						(:line 0 :character 0)
						:end
						(:line 0 :character 0))
					 :context
					 (:diagnostics
						[]
						:only
						["source.organizeImports"])))
[server-reply] (id:3) Fri Jan 29 13:06:03 2021:
(:jsonrpc "2.0" :result
					[(:title "Organize Imports" :kind "source.organizeImports" :disabled
									 (:reason "")
									 :edit
									 (:documentChanges
										[(:textDocument
											(:version 0 :uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
											:edits
											[(:range
												(:start
												 (:line 1 :character 0)
												 :end
												 (:line 1 :character 0))
												:newText "\nimport \"golang.org/x/mod/module\"\n")])]))]
					:id 3)
[client-notification] Fri Jan 29 13:06:03 2021:
(:jsonrpc "2.0" :method "textDocument/didChange" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 1)
					 :contentChanges
					 [(:range
						 (:start
							(:line 1 :character 0)
							:end
							(:line 1 :character 0))
						 :rangeLength 0 :text "\nimport \"golang.org/x/mod/module\"\n")]))
[server-notification] Fri Jan 29 13:06:03 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
					(:type 3 :message "2021/01/29 13:06:03 go/packages.Load\n	snapshot=2\n	directory=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411\n	query=[file=/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go]\n	packages=1\n"))
[server-notification] Fri Jan 29 13:06:03 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
					(:type 3 :message "2021/01/29 13:06:03 go/packages.Load\n	snapshot=2\n	package=\"example.com\"\n	files=[/usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go]\n"))
[server-notification] Fri Jan 29 13:06:03 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
					(:type 1 :message "2021/01/29 13:06:03 golang.org/x/mod/module: no dep handle for golang.org/x/xerrors: no metadata for golang.org/x/xerrors\n	snapshot=2\n"))
[server-notification] Fri Jan 29 13:06:03 2021:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
					(:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 1 :diagnostics
								[]))
[client-notification] Fri Jan 29 13:06:06 2021:
(:jsonrpc "2.0" :method "textDocument/didChange" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 2)
					 :contentChanges
					 [(:range
						 (:start
							(:line 2 :character 0)
							:end
							(:line 3 :character 0))
						 :rangeLength 33 :text "")]))
[server-notification] Fri Jan 29 13:06:06 2021:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
					(:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 2 :diagnostics
								[(:range
									(:start
									 (:line 4 :character 16)
									 :end
									 (:line 4 :character 22))
									:severity 1 :source "compiler" :message "undeclared name: module")]))
[internal] (id:4) Fri Jan 29 13:06:07 2021:
(:deferring :textDocument/signatureHelp :id 4 :params
						(:textDocument
						 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
						 :position
						 (:line 0 :character 0)))
[internal] (id:5) Fri Jan 29 13:06:07 2021:
(:deferring :textDocument/hover :id 5 :params
						(:textDocument
						 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
						 :position
						 (:line 0 :character 0)))
[internal] (id:6) Fri Jan 29 13:06:07 2021:
(:deferring :textDocument/documentHighlight :id 6 :params
						(:textDocument
						 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
						 :position
						 (:line 0 :character 0)))
[client-notification] Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :method "textDocument/didChange" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 3)
					 :contentChanges
					 [(:range
						 (:start
							(:line 1 :character 0)
							:end
							(:line 2 :character 0))
						 :rangeLength 1 :text "")]))
[internal] Fri Jan 29 13:06:07 2021:
(:maybe-run-deferred
 (4 5 6))
[client-request] (id:4) Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :id 4 :method "textDocument/signatureHelp" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
					 :position
					 (:line 0 :character 0)))
[client-request] (id:5) Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :id 5 :method "textDocument/hover" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
					 :position
					 (:line 0 :character 0)))
[client-request] (id:6) Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :id 6 :method "textDocument/documentHighlight" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
					 :position
					 (:line 0 :character 0)))
[server-notification] Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :method "window/logMessage" :params
					(:type 1 :message "2021/01/29 13:06:07 no signature help: cannot find an enclosing function\n	position={0 0}\n"))
[server-reply] (id:4) Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :result nil :id 4)
[server-reply] (id:5) Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :result nil :id 5)
[server-notification] Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
					(:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 3 :diagnostics
								[(:range
									(:start
									 (:line 3 :character 16)
									 :end
									 (:line 3 :character 22))
									:severity 1 :source "compiler" :message "undeclared name: module")]))
[server-reply] (id:6) Fri Jan 29 13:06:07 2021:
(:jsonrpc "2.0" :result
					[]
					:id 6)
[client-request] (id:7) Fri Jan 29 13:06:08 2021:
(:jsonrpc "2.0" :id 7 :method "textDocument/codeAction" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
					 :range
					 (:start
						(:line 0 :character 2)
						:end
						(:line 0 :character 2))
					 :context
					 (:diagnostics
						[]
						:only
						["source.organizeImports"])))
[server-reply] (id:7) Fri Jan 29 13:06:08 2021:
(:jsonrpc "2.0" :result
					[(:title "Organize Imports" :kind "source.organizeImports" :disabled
									 (:reason "")
									 :edit
									 (:documentChanges
										[(:textDocument
											(:version 3 :uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
											:edits
											[(:range
												(:start
												 (:line 1 :character 0)
												 :end
												 (:line 1 :character 0))
												:newText "\nimport \"golang.org/x/mod/module\"\n")])]))]
					:id 7)
[client-notification] Fri Jan 29 13:06:09 2021:
(:jsonrpc "2.0" :method "textDocument/didChange" :params
					(:textDocument
					 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 5)
					 :contentChanges
					 [(:range
						 (:start
							(:line 0 :character 0)
							:end
							(:line 0 :character 0))
						 :rangeLength 0 :text " ")
						(:range
						 (:start
							(:line 0 :character 1)
							:end
							(:line 0 :character 1))
						 :rangeLength 0 :text " ")]))
[internal] (id:4) Fri Jan 29 13:06:17 2021:
(:timed-out :textDocument/signatureHelp :id 4 :params
						(:textDocument
						 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
						 :position
						 (:line 0 :character 0)))
[internal] (id:5) Fri Jan 29 13:06:17 2021:
(:timed-out :textDocument/hover :id 5 :params
						(:textDocument
						 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
						 :position
						 (:line 0 :character 0)))
[internal] (id:6) Fri Jan 29 13:06:17 2021:
(:timed-out :textDocument/documentHighlight :id 6 :params
						(:textDocument
						 (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
						 :position
						 (:line 0 :character 0)))
```

**Comment by joaotavora** [comment]
Yes, I think I see the problem.  You're asking for code actions before Eglot has a chance to tell the server that it recently made changes to the buffer.  So all the actions you get are stale when it's time to actually run them.  Thank the server for figuring this out.

I'll have to think about how this can be fixed.

**Comment by muffinmad** [comment]
@joaotavora 
Can it be the timer and the `eglot--versioned-identifier` issue in the `eglot--after-change`?

My guess from this two events:

```
server-reply (id:2) Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :result
	  [(:title "Organize Imports" :kind "source.organizeImports" :disabled
		   (:reason "")
		   :edit
		   (:documentChanges
		    [(:textDocument
		      (:version 0 :uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go")
		      :edits
		      [(:range
			(:start
			 (:line 0 :character 0)
			 :end
			 (:line 0 :character 2))
			:newText "")
		       (:range
			(:start
			 (:line 1 :character 0)
			 :end
			 (:line 1 :character 0))
			:newText "\nimport \"golang.org/x/mod/module\"\n")])]))]
	  :id 2)

client-notification Fri Jan 29 12:35:04 2021:
(:jsonrpc "2.0" :method "textDocument/didChange" :params
	  (:textDocument
	   (:uri "file:///usr/local/google/home/bcmills/src/github.com/bcmills/eglot-issue-411/main.go" :version 1)
	   :contentChanges
	   [(:range
	     (:start
	      (:line 0 :character 0)
	      :end
	      (:line 0 :character 0))
	     :rangeLength 0 :text " ")]))
```

1. Edit occur and the timer is activated in `eglot--after-change` to send `didChange` later. Server thinks that document version is 0
2. `codeActions` request is made and the server response with TextEdit for document version 0.
3. Eglot apply the code action and check for version by `(equal version eglot--versioned-identifier)`

And the problem is that `eglot--versioned-identifier` is incremented outside the timer. Eglot think that document is version 1 after step 1 and before server is notified about changes.

**Comment by joaotavora** [comment]
It might be that yes, @muffinmad .  I thought about this too, but I'm not 100% sure if it would make a difference. Try it out.  It should be easy enough, and perhaps more correct, to increment the "official" document version _exactly_ when we tell the server that the document changed, i.e. in the `didChange`.

have fun :-)

**Comment by muffinmad** [comment]
But we can't apply the TextEdit after document is changed, right? TextEdit is about to replace the whole first line with so the text entered by user will be lost.

Can the pending `didChange` events be pushed to the server before `codeActions` request?

**Comment by joaotavora** [comment]
> Can the pending didChange events be pushed to the server before codeActions request?

Likely that's the way it needs to work yes. And perhaps not only codeActions, but also other things that involve an edit.  I think the problem with code actions is more severe though, and maybe the only worth fixing.  because you first get a list of actions, then potentially a lot of time elapses, then you select the action.

**Comment by muffinmad** [comment]
Wait, shouldn't it work like that already?

```
jsonrpc-request is a compiled Lisp function in ‘jsonrpc.el’.

This function has :before advice: eglot--signal-textDocument/didChange
```

**Comment by joaotavora** [comment]
huuhh, you're right.  you found my dirty little secret i'd forgotten about :-D  So something else must be happening.

**Comment by muffinmad** [comment]
> because you first get a list of actions, then potentially a lot of time elapses, then you select the action.

Do not forget that minibuffer is not modal. One can get the list of actions, switch to buffer, change everything, switch to minibuffer and select some action as if nothing had happened :)

**Comment by joaotavora** [comment]
> Do not forget that minibuffer is not modal.

It can be though, I think, there's an option for that.  I saw a discussion on emacs-devel about that recently.  

Anyway if the user is doing that he's asking for the error message anyway, so I wouldn't worry much about that.

**Comment by muffinmad** [comment]
> huuhh, you're right. you found my dirty little secret i'd forgotten about :-D So something else must be happening.

But `jsonrpc-request` for `codeActions` is performed without the `:deferred`. Correct me if I wrong, advice is calling the `eglot--signal-textDocument/didChange` function only when `(and eglot--managed-mode deferred)`.

**Comment by joaotavora** [comment]
Ah it might be that yes.  So either we use `:deferred` in `eglot-code-actions` or we explicitly send the didChange before sending stuff.  At least it's something to experiment with.  I don't have much time right now, but anyway this is really good analysis/investigation, so thanks!

**Comment by muffinmad** [comment]
@bcmills Please try this little patch

```diff
diff --git a/eglot.el b/eglot.el
index 98e0bcf..8403e5d 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2554,7 +2554,8 @@ at point.  With prefix argument, prompt for ACTION-KIND."
                                when (cdr (assoc 'eglot-lsp-diag
                                                 (eglot--diag-data diag)))
                                collect it)]
-                   ,@(when action-kind `(:only [,action-kind]))))))
+                   ,@(when action-kind `(:only [,action-kind]))))
+           :deferred t))
          (menu-items
           (or (cl-loop for action across actions
                        ;; Do filtering ourselves, in case the `:only'
```

**Comment by bcmills** [comment]
@muffinmad, that certainly seems to improve things, at least.

(It at least fixes the script in my reproducer repo, and in limited testing it seems to fix manually-triggered actions too. 👍)
